LFLT-557 Implement the log collector logic in TLC

petersmith-hun opened the pull request:
- Implemented log collection pipeline
 - Implemented Docker Engine API listener
 - Implemented byte array and joining json parsers
 - Implemented identity, Logstash JSON and custom JSON log object mapping
 - Implemented console and TLP publisher
 - Implemented pipeline controller and stream reconnection logic
 - Added default configuration

petersmith-hun committed:
LFLT-557 Implement the log collector logic in TLC
 - Implemented log collection pipeline
 - Implemented Docker Engine API listener
 - Implemented byte array and joining json parsers
 - Implemented identity, Logstash JSON and custom JSON log object mapping
 - Implemented console and TLP publisher
 - Implemented pipeline controller and stream reconnection logic
 - Added default configuration

elementbound commented on src/tlc/client/docker/docker-engine-api-client.ts:
```diff
@@ -0,0 +1,86 @@
+    private readonly dockerConnection: DockerConnection;
+    private readonly systemConfig: SystemConfig;
+
+    private constructor() {
```
Dependencies not passed through constructor args? How controversial :smile:

petersmith-hun replied:
Don't even start...! :D

elementbound commented on src/tlc/client/docker/docker-engine-api-client.ts:
```diff
@@ -0,0 +1,86 @@
+    /**
+     * Returns a singleton instance of the DockerEngineApiClient.
+     */
+    public static getInstance(): DockerEngineApiClient {
```
What do you think about the pros and cons of explicitly implementing a singleton pattern instead of just exporting a const variable that holds the instance?

petersmith-hun replied:
Yeah, already done.

elementbound commented on src/tlc/client/docker/docker-engine-api-client.ts:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    /**
+     * Retrieves the list of existing containers (regardless their status).
```
typo: regardless _of_

elementbound commented on src/tlc/client/docker/index.ts:
```diff
@@ -0,0 +1,7 @@
+ * Data interface for Docker container definitions.
+ */
+export interface ContainerDefinition {
+    Id: string;
```
Minor, but why the capitalized property names?

petersmith-hun replied:
Docker Engine API .. :( It exposes the properties like this...

elementbound commented on src/tlc/config/configuration-provider.ts:
```diff
@@ -0,0 +1,65 @@
+     * Refers to the parameters under {@code tlc.system} section.
+     * @see SystemConfig
+     */
+    getSystemConfig(): SystemConfig {
```
What do you think about using [getters](https://www.typescriptlang.org/docs/handbook/classes.html#accessors)?

elementbound commented on src/tlc/controller.ts:
```diff
@@ -0,0 +1,62 @@
+
+    /**
+     * Initializes all pipelines. It does the following steps:
+     *  1) Creates the disconnection subject to passed to the pipelines;
```
Typo? Creates the disconnection subject _to pass_ to the pipelines

petersmith-hun replied:
It was meant to be "to be passed", but yeah, typo.

elementbound commented on src/tlc/pipeline/listener/docker-logs-api-listener.ts:
```diff
@@ -0,0 +1,64 @@
+
+    private initListener(): Promise<AxiosResponse<IncomingMessage>> {
+
+        return this.dockerEngineApiClient.getContainers()
```
Is there any reason for not just writing an async method?

petersmith-hun replied:
The entire thing is wired together using Rx, that gives more control over the data flow. But I'm not sure what you mean for this specific line, care to elaborate?

elementbound replied:
getContainers returns a Promise, which then you chain with .then calls - I was wondering why not just go the usual async/await route

petersmith-hun replied:
The entire thing is async anyways, so I figured it's okay like this. As you can see, even if I awaited the promise, the `incomingMessage.on("data")` calls are still pushing out the messages via the observables in an async manner.

elementbound commented on src/tlc/pipeline/parser/joining-json-parser.ts:
```diff
@@ -0,0 +1,47 @@
+            if (inputData.endsWith(newLineCharacter)) {
+                this.buffer.join("")
+                    .trim()
+                    .split(newLineCharacter)
```
Nitpick: what's the benefit of a constant here instead of writing "\n"?

petersmith-hun replied:
Avoiding repetition (line #32 and #35).

elementbound commented on src/tlc/pipeline/parser/joining-json-parser.ts:
```diff
@@ -0,0 +1,47 @@
+            this.buffer.push(inputData);
+            if (inputData.endsWith(newLineCharacter)) {
+                this.buffer.join("")
+                    .trim()
```
I'm not sure if that's intended, but I've just figured out that trimming has an additional benefit - if the input ends with a "\n", then the split will produce an empty string at the end of the array. However, since it's trimmed, it won't.

petersmith-hun replied:
Yep, I used a similar logic in Domino for processing streaming data from Docker Engine, and it was needed for some reason. I'm not exactly sure anymore why, but I think the data lines were having some extra white spaces, that needed to be removed. Same source, same precautions. :)

elementbound commented on src/tlc/pipeline/parser/joining-json-parser.ts:
```diff
@@ -0,0 +1,47 @@
+                        try {
+                            subscriber.next(JSON.parse(value));
+                        } catch (error) {
+                            log.warn("Log message could not be parsed as JSON");
```
Is it possible for subscriber.next to throw? Because in that case you might be accidentally hiding errors.

petersmith-hun replied:
It will just pass the data to the next observer, so on its own, it is not supposed to be throwing errors. The pipeline is more-or-less foolproof, but I actually expect some issues from edge cases to reveal themeselves after a while. Hope that's not actually gonna happen.

petersmith-hun committed:
LFLT-557 Implement the log collector logic in TLC
 - Removed singleton providers

petersmith-hun committed:
LFLT-557 Implement the log collector logic in TLC
 - Fixing PR comments

petersmith-hun committed:
LFLT-557 Implement the log collector logic in TLC
 - Dependency handling refactor

petersmith-hun committed:
LFLT-557 Implement the log collector logic in TLC
 - Import fixes
 - Fixed MapperFactory (it was returning the same instance of mappers)
 - Added unit tests
 - Configured collecting coverage info

petersmith-hun committed:
LFLT-557 Implement the log collector logic in TLC
 - Bumped version to 1.0.0-dev

petersmith-hun committed:
LFLT-557 Implement the log collector logic in TLC
 - Fixed high CPU usage while executing tests
